rskj-core/src/main/java/co/rsk/core/bc/GarbageCollector.java

```diff
@@ -0,0 +1,91 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.core.bc;
+
+import co.rsk.config.InternalService;
+import co.rsk.db.RepositoryLocator;
+import co.rsk.trie.MultiTrieStore;
+import org.ethereum.core.Block;
+import org.ethereum.core.BlockHeader;
+import org.ethereum.core.TransactionReceipt;
+import org.ethereum.db.BlockStore;
+import org.ethereum.listener.CompositeEthereumListener;
+import org.ethereum.listener.EthereumListener;
+import org.ethereum.listener.EthereumListenerAdapter;
+
+import java.util.List;
+
+public class GarbageCollector implements InternalService {
+    private final CompositeEthereumListener emitter;
+    private final MultiTrieStore multiTrieStore;
+    private final BlockStore blockStore;
+    private final RepositoryLocator repositoryLocator;
+    private final EthereumListener garbageCollectorInvoker;
+
+    public GarbageCollector(CompositeEthereumListener emitter,
+                            int blocksPerEpoch,
+                            int numberOfEpochs,
+                            MultiTrieStore multiTrieStore,
+                            BlockStore blockStore,
+                            RepositoryLocator repositoryLocator) {
+        this.emitter = emitter;
+        this.multiTrieStore = multiTrieStore;
+        this.blockStore = blockStore;
+        this.repositoryLocator = repositoryLocator;
+        this.garbageCollectorInvoker = new GarbageCollectorInvoker(blocksPerEpoch, numberOfEpochs);
+    }
+
+    @Override
+    public void start() {
+        emitter.addListener(garbageCollectorInvoker);
+    }
+
+    @Override
+    public void stop() {
+        emitter.removeListener(garbageCollectorInvoker);
+    }
+
+    private void collect(long untilBlock) {
+        BlockHeader untilHeader = blockStore.getChainBlockByNumber(untilBlock).getHeader();
+        multiTrieStore.collect(repositoryLocator.snapshotAt(untilHeader).getRoot());
+    }
+
+    private class GarbageCollectorInvoker extends EthereumListenerAdapter {
+        private final int blocksPerEpoch;
+        private final int numberOfEpochs;
+
+        public GarbageCollectorInvoker(int blocksPerEpoch, int numberOfEpochs) {
+            this.blocksPerEpoch = blocksPerEpoch;
+            this.numberOfEpochs = numberOfEpochs;
+        }
+
+        /**
+         * It'll collect all states older than <code>blocksPerEpoch * numberOfEpochs - 1</code> from
+         * the store (if there are enough blocks)
+         */
+        @Override
+        public void onBestBlock(Block block, List<TransactionReceipt> receipts) {
+            int statesToKeep = blocksPerEpoch * numberOfEpochs - 1;
+            long currentBlockNumber = block.getNumber();
+            if(currentBlockNumber % blocksPerEpoch == 0 && currentBlockNumber >= statesToKeep) {
+                collect(currentBlockNumber - statesToKeep);
```

Comment:
I find it strange that this logic is deciding statesToKeep and it has to match MultiTrieStore collect's save into the upcoming last epoch

Reply:
I'm not quite sure about this. I think passing the latest accesible hash after the `collect` takes place is independent on where that trie is stored